### PR TITLE
Implement the new Symfony 6.x security interfaces

### PR DESCRIPTION
**Pull request by bytehead**
- [x] Backend
- [x] Frontend
- [x] Preview (Anonymous)
- [x] Preview (Member)
- [x] Remember-Me
- [x] Backend 2FA
- [x] Frontend 2FA
- [x] Logout
- [x] Tests
- [x] Remove `importUser` hook (see #4973)

**Issues:**
- [x] Backup codes are not working for me ATM. I can generate them, but when I try to use them, Contao says "Invalid verification code". (Fixed by config in 4b10edbcb94ecd98d3256a3fc0c80aa57487c2e4)
- [x] I can add my browser as a trusted device, however, when I log in again, Contao still asks for a verification code. (Fixed by config in bbfb29a8cf)
- [x] In the front end preview, I can log in a particular user but I can no longer log them out again by leaving the "username" field blank. The user stays logged in. (Fixed in 4c3ba9c621)

**TODO**
- [ ] Adjust `contao/docs` about 4b10edbcb94ecd98d3256a3fc0c80aa57487c2e4.
- [ ] Adjust `contao/docs` about bbfb29a8cf.

### CODE REVIEW

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        /** @var PageError401 $pageHandler */
+        $pageHandler = new $GLOBALS['TL_PTY']['error_401']();
+
+        try {
+            return $pageHandler->getResponse();
+        } catch (ResponseException $e) {
+            return $e->getResponse();
+        } catch (InsufficientAuthenticationException $e) {
+            throw new UnauthorizedHttpException('', $e->getMessage(), $e);
+        }
```
Comment (aschempp) [inline core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: should always throw an exception, our PrettyErrorScreenListener will take care of rendering the login screen

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        return $request->isMethod('POST')
+            && $request->request->has('FORM_SUBMIT')
+            && 0 === strncmp($request->request->get('FORM_SUBMIT'), 'tl_login', 8);
```
Comment (aschempp) [inline core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: didn't we recently change something in this handling? /cc @fritzmg

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        if ($currentToken instanceof TwoFactorTokenInterface) {
+            return $this->twoFactorAuthenticator->authenticate($request);
+        }
+
+        $credentials = $this->getCredentials($request);
+
+        $passport = new Passport(
+            new UserBadge($credentials['username'], [$this->userProvider, 'loadUserByIdentifier']),
+            new PasswordCredentials($credentials['password']),
+            [new RememberMeBadge()]
```
Comment (aschempp) [inline core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php:127]: Should this maybe only be added if the checkbox is checked in the front end?

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    {
+        return $token->allTwoFactorProvidersAuthenticated();
+    }
+
+    private function getCredentials(Request $request): array
+    {
+        $credentials = [];
+        $credentials['csrf_token'] = ParameterBagUtils::getRequestParameterValue($request, $this->options['csrf_parameter']);
+
+        if ($this->options['post_only']) {
```
Comment (aschempp) [inline core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: I don't think we intentionally want to support any other options

File: core-bundle/src/Security/User/ContaoUserProvider.php [hunk 5]
```diff
@@ -58,8 +58,6 @@ public function __construct(ContaoFramework $framework, SessionInterface $sessio
      */
     public function loadUserByUsername($username): User
     {
-        trigger_deprecation('contao/core-bundle', '4.13', 'Using "ContaoUserProvider::loadUserByUsername()" has been deprecated and will no longer work in Contao 5.0. Use "ContaoUserProvider::loadUserByIdentifier()" instead.');
```
Comment (aschempp) [inline core-bundle/src/Security/User/ContaoUserProvider.php:48]: Why is this no longer deprecated? Ans why does the method still have the annotation?

**Comment by aschempp** [comment]
ping @bytehead anything I can help to make some progress here? 🙃

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 1 above)
Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: Adjusted in c15ed5f8f.

File: core-bundle/src/Security/User/ContaoUserProvider.php (hunk 5 above)
Comment (bytehead) [reply core-bundle/src/Security/User/ContaoUserProvider.php:48]: As long as we support Symfony 5.x we will have to implement the method. It's deprecated since 5.3 and removed in 6.0. Don't know how to proceed from here?

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 4 above)
Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: Removed in bbd5c520d.

File: core-bundle/src/Security/User/ContaoUserProvider.php (hunk 5 above)
Comment (bytehead) [reply core-bundle/src/Security/User/ContaoUserProvider.php:48]: May we also deprecate it in 5.x and drop it in 6.0 as we support Symfony 6.x only then?

File: core-bundle/composer.json [hunk 6]
```diff
@@ -126,8 +126,8 @@
         "symfony/process": "^5.4",
         "symfony/property-access": "^5.4",
         "symfony/routing": "^5.4",
-        "symfony/security-bundle": "^5.4",
-        "symfony/security-core": "^5.4",
+        "symfony/security-bundle": "^6.0",
```
Comment (ausi) [inline core-bundle/composer.json]: This needs to be changed in *manager-bundle/composer.json* too.

Comment (bytehead) [reply core-bundle/composer.json]: 👍 fixed in a2c63ac1cf.

File: core-bundle/src/Security/User/ContaoUserProvider.php (hunk 5 above)
Comment (bytehead) [reply core-bundle/src/Security/User/ContaoUserProvider.php:48]: I'm gonna use Symfony 6.x Security only (see 23d36e53cb3b79f8a561071b9652007427f645eb).

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 2 above)
Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: ping @aschempp @fritzmg what does have changed here? :)

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 3 above)
Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php:127]: > Without this badge, remember me will be not be activated (regardless of all other settings).

https://symfony.com/doc/current/security/remember_me.html#add-remember-me-support-to-the-authenticator

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 2 above)
Comment (fritzmg) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: Yes, see https://github.com/contao/contao/pull/3981. (This change is also present in `5.x` https://github.com/contao/contao/blob/5.x/core-bundle/src/Security/Authentication/ContaoLoginAuthenticationListener.php#L51)

Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: Thanks, I'll check.

Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: Fixed in 7d4afa7071.

**Comment by bytehead** [comment]
Please review roughly, before I add the missing tests 😬

File: core-bundle/src/EventListener/PreviewAuthenticationListener.php [hunk 7]
```diff
@@ -43,7 +43,7 @@ public function __invoke(RequestEvent $event): void
             !$request->attributes->get('_preview', false)
             || $this->scopeMatcher->isBackendRequest($request)
             || $this->tokenChecker->hasBackendUser()
-            || $this->security->getToken() instanceof FrontendPreviewToken
+            || $this->security->getToken() instanceof UsernamePasswordToken
```
Comment (aschempp) [inline core-bundle/src/EventListener/PreviewAuthenticationListener.php]: Not sure this works correctly. How do I now detect a front end preview without username and without "unpublished elements"? I think this will break preview links

File: core-bundle/src/EventListener/Security/LogoutListener.php [hunk 8]
```diff
@@ -90,27 +87,6 @@ private function logout(Request $request): void
                 sprintf('User "%s" has logged out', $user->username),
                 ['contao' => new ContaoContext(__METHOD__, ContaoContext::ACCESS, $user->username)]
             );
-
-            $this->triggerPostLogoutHook($user);
```
Comment (aschempp) [inline core-bundle/src/EventListener/Security/LogoutListener.php:94]: we should add to the readme that the `postLogout` hook is gone

File: core-bundle/src/Resources/contao/library/Contao/User.php [hunk 9]
```diff
@@ -392,14 +380,9 @@ public static function loadUserByIdentifier(string $identifier): self|null
 
 	/**
 	 * {@inheritdoc}
-	 *
-	 * @deprecated Deprecated since Contao 4.13, to be removed in Contao 5.0.
-	 *             Use Contao\User::getUserIdentifier() instead.
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/User.php:397]: why did you remove this? The method seems gone in Sf? https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/Security/Core/User/UserInterface.php

File: core-bundle/src/Security/Authentication/FrontendPreviewAuthenticator.php [hunk 10]
```diff
@@ -43,18 +43,17 @@ public function authenticateFrontendUser(string $username, bool $showUnpublished
             return false;
         }
 
-        $token = new FrontendPreviewToken($user, $showUnpublished);
+        $token = new UsernamePasswordToken($user, 'contao_frontend');
 
         $this->session->set('_security_contao_frontend', serialize($token));
+        $this->session->set('_contao_frontend_preview', ['showUnpublished' => $showUnpublished]);
```
Comment (aschempp) [inline core-bundle/src/Security/Authentication/FrontendPreviewAuthenticator.php]: we need to make sure to remove `_contao_frontend_preview` from the session on user logout!

File: core-bundle/src/Security/Authentication/Token/TokenChecker.php [hunk 11]
```diff
@@ -101,9 +101,15 @@ public function isPreviewMode(): bool
             return false;
         }
 
-        $token = $this->getToken(self::FRONTEND_FIREWALL);
+        $session = $request->getSession();
```
Comment (aschempp) [inline core-bundle/src/Security/Authentication/Token/TokenChecker.php:117]: what if the request does not have a session? Maybe check `$request->hasPreviousSession()` above?

File: core-bundle/src/Security/Authentication/Token/TokenChecker.php [hunk 12]
```diff
@@ -114,11 +120,11 @@ private function getToken(string $context): TokenInterface|null
             $token = $this->getTokenFromSession('_security_'.$context);
         }
 
-        if (!$token instanceof TokenInterface || !$token->isAuthenticated()) {
+        if (!$token instanceof TokenInterface) {
             return null;
         }
 
-        if ($this->trustResolver->isAnonymous($token)) {
+        if (!$this->trustResolver->isAuthenticated($token) || !$this->trustResolver->isFullFledged($token)) {
```
Comment (aschempp) [inline core-bundle/src/Security/Authentication/Token/TokenChecker.php]: `isFullFledged` would mean a rememberme token no longer qualifies here? Is that intentional?

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+        if ($this->scopeMatcher->isBackendRequest($request)) {
+            return $this->redirectToBackend($request);
+        }
+
+        $this->framework->initialize();
+
+        $page = $request->attributes->get('pageModel');
+
+        if (!$page instanceof PageModel) {
+            throw new \RuntimeException('No PageModel found on request.');
```
Comment (aschempp) [inline core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: Not sure we can throw an exception here. It is valid to have a controller without a page model inside the front end firewall. Not sure what to do here then though 😬. Maybe not return `true` on `isInteractive`?

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+        }
+
+        $page->loadDetails();
+        $page->protected = false;
+
+        $route = $this->pageRegistry->getRoute($page);
+        $subRequest = $request->duplicate(null, null, $route->getDefaults());
```
Comment (aschempp) [inline core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php:96]: Looks like this is all unused if the token storage does not have a token?

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+        }
+
+        return $passport;
+    }
+
+    public function createToken(Passport $passport, string $firewallName): TokenInterface
+    {
+        $credentialsBadge = $passport->getBadge(TwoFactorCodeCredentials::class);
+
+        if ($credentialsBadge instanceof TwoFactorCodeCredentials) {
```
Comment (aschempp) [inline core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: might want to early return here instead with parent::createToken

File: core-bundle/src/Security/User/ContaoUserProvider.php [hunk 16]
```diff
@@ -98,19 +81,4 @@ public function upgradePassword(UserInterface $user, string $newHashedPassword):
         $user->password = $newHashedPassword;
         $user->save();
     }
-
-    private function triggerPostAuthenticateHook(User $user): void
-    {
-        if (empty($GLOBALS['TL_HOOKS']['postAuthenticate']) || !\is_array($GLOBALS['TL_HOOKS']['postAuthenticate'])) {
```
Comment (aschempp) [inline core-bundle/src/Security/User/ContaoUserProvider.php:104]: should be added to the `UPGRADE.md`

File: core-bundle/tests/Functional/app/config/security.yaml [hunk 17]
```diff
@@ -1,14 +1,16 @@
 security:
+    enable_authenticator_manager: true
```
Comment (aschempp) [inline core-bundle/tests/Functional/app/config/security.yaml]: I thought this is enabled in Sf 6 by default/cannot be disabled?

File: core-bundle/tests/Functional/app/config/security.yaml [hunk 18]
```diff
@@ -1,14 +1,16 @@
...
+
+    password_hashers:
+        Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface: auto
```
Comment (aschempp) [inline core-bundle/tests/Functional/app/config/security.yaml]: is this change necessary? Looks like it would force this on all users in the application, not just Contao user objects?

File: core-bundle/src/Resources/contao/library/Contao/User.php (hunk 9 above)
Comment (bytehead) [reply core-bundle/src/Resources/contao/library/Contao/User.php:397]: Leftover, removed all of them in 2c6916e170.

File: core-bundle/tests/Functional/app/config/security.yaml (hunk 17 above)
Comment (bytehead) [reply core-bundle/tests/Functional/app/config/security.yaml]: Correct, is not needed anymore. Removed in b2814d45626ce8e7360470ba4a34e36b2b2a7991.

File: core-bundle/tests/Functional/app/config/security.yaml (hunk 18 above)
Comment (bytehead) [reply core-bundle/tests/Functional/app/config/security.yaml]: We should only apply this to `Contao\User` objects. Fixed in b0669adc4b.

File: core-bundle/tests/Functional/app/config/security.yaml (hunk 17 above)
Comment (bytehead) [reply core-bundle/tests/Functional/app/config/security.yaml]: I'll have to revert: https://github.com/symfony/security-bundle/blob/6.1/DependencyInjection/SecurityExtension.php#L97

Comment (bytehead) [reply core-bundle/tests/Functional/app/config/security.yaml]: Reverted in 099c9a4b69.

Comment (bytehead) [reply core-bundle/tests/Functional/app/config/security.yaml]: Forget all about it - it was buggy in 6.0.0 and 6.0.1. Need to require 6.0.2 (fixed in 01a42c5228).

File: core-bundle/src/EventListener/Security/LogoutListener.php (hunk 8 above)
Comment (bytehead) [reply core-bundle/src/EventListener/Security/LogoutListener.php:94]: Added in f178248ae1.

File: core-bundle/src/Security/User/ContaoUserProvider.php (hunk 16 above)
Comment (bytehead) [reply core-bundle/src/Security/User/ContaoUserProvider.php:104]: Added in 94145d9fc8.

File: core-bundle/src/Security/Authentication/FrontendPreviewAuthenticator.php (hunk 10 above)
Comment (bytehead) [reply core-bundle/src/Security/Authentication/FrontendPreviewAuthenticator.php]: Added in 045a109e4a.

File: core-bundle/src/Security/Authentication/Token/TokenChecker.php (hunk 11 above)
Comment (bytehead) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php:117]: Done in f281ddfe26.

File: core-bundle/src/EventListener/PreviewAuthenticationListener.php (hunk 7 above)
Comment (bytehead) [reply core-bundle/src/EventListener/PreviewAuthenticationListener.php]: Why would it break?

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 15 above)
Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: Changed in 68fc2389dd.

File: core-bundle/src/Security/Authentication/Token/TokenChecker.php (hunk 12 above)
Comment (bytehead) [reply core-bundle/src/Security/Authentication/Token/TokenChecker.php]: Added `isRememberMe` in 8582db1822.

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 14 above)
Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php:96]: Yes, I tried to avoid the `else` clause.

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 13 above)
Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: Could be possible for this case. But what about the exception, if there is no error page configured (see line 106, https://github.com/contao/contao/pull/4343/files#diff-52c9f08efd8ad0636bb94ec00aca18c33b660927772cb2a25c26c2c92f7b0fc1R106)?

**Comment by xprojects-de** [comment]
Servus zusammen,

erwarte sehensüchtig den merge dieses PRs :-)
Wann ist es denn soweit ? :-)

Danke und Grüße

File: core-bundle/src/EventListener/PreviewAuthenticationListener.php (hunk 7 above)
Comment (bytehead) [reply core-bundle/src/EventListener/PreviewAuthenticationListener.php]: Adjusted in 774e7781b1.

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 13 above)
Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php]: Adjusted in d2d9bdc98b.

If there is no 401 page configured, it will show an UnauthorizedHttpException:

Bildschirmfoto 2022-07-06 um 16 53 31

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+
+        return true;
+    }
+
+    private function getCredentials(Request $request): array
+    {
+        $credentials = [
+            'csrf_token' => ParameterBagUtils::getRequestParameterValue($request, $this->options['csrf_parameter']),
+            'username' => ParameterBagUtils::getParameterBagValue($request->request, $this->options['username_parameter']),
+            'password' => ParameterBagUtils::getParameterBagValue($request->request, $this->options['password_parameter']) ?? '',
```
Comment (leofeyer) [inline core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php:193]: I assume there is no other way, is there?

**Review by leofeyer — COMMENTED** [review]
The PR is looking very good. Great job @bytehead!

I have noticed that the (deprecated) `postLogin` hook is still present although the `postLogout` and some other hooks have been removed. Is this intentional?

**Review by leofeyer — CHANGES_REQUESTED** [review]
I have found a few show stoppers, I‘m afraid. 🙈 

1. Backup codes are not working for me ATM. I can generate them, but when I try to use them, Contao says "Invalid verification code".

2. I can add my browser as a trusted device, however, when I log in again, Contao still asks for a verification code.

3. In the front end preview, I can log in a particular user but I can no longer log them out again by leaving the "username" field blank. The user stays logged in.

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 19 above)
Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php:193]: I don't get it why this should be internal?

File: core-bundle/src/Security/Authentication/RememberMe/RememberMeTokenProvider.php [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+use Symfony\Component\Security\Core\Authentication\RememberMe\PersistentTokenInterface;
+use Symfony\Component\Security\Core\Authentication\RememberMe\TokenProviderInterface;
+
+class RememberMeTokenProvider implements TokenProviderInterface
+{
+    public function __construct(private RememberMeRepository $repository)
+    {
+    }
+
+    public function loadTokenBySeries(string $series): PersistentToken
```
[the remark this answers is not in the document]
Comment (bytehead) [reply core-bundle/src/Security/Authentication/RememberMe/RememberMeTokenProvider.php:27]: Same here? 🤷

**Comment by bytehead** [comment]
@leofeyer Thanks a lot for testing, I'll address the issues!

File: core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php (hunk 19 above)
Comment (bytehead) [reply core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php:193]: As discussed with @leofeyer we will keep it as it is and tackle/discuss the `@internal` usages all over Contao in a different PR.

File: core-bundle/src/Security/Authentication/RememberMe/RememberMeTokenProvider.php (hunk 20 above)
[the remark this answers is not in the document]
Comment (bytehead) [reply core-bundle/src/Security/Authentication/RememberMe/RememberMeTokenProvider.php:27]: As discussed with @leofeyer we will keep it as it is and tackle/discuss the `@internal` usages all over Contao in a different PR.

**Comment by bytehead** [comment]
@leofeyer please test again and check your config, you'll have to configure the custom managers:

```
scheb_two_factor:
    trusted_device:
        enabled: true
        manager: contao.security.two_factor.trusted_device_manager
    backup_codes:
        enabled: true
        manager: contao.security.two_factor.backup_code_manager
```

**Comment by bytehead** [comment]
Removed the `postLogin` hook in f0e97964cf.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bytehead.
